Make text fields yield keyboard focus on "Esc" in a consistent way

dragunoff opened the pull request:
Some text fields were clearing the input on <kbd>Esc</kbd> before yielding focus and others were only clearing input (lobby chat field) and others were doing both at the same time (map editor search filter). This PR standardizes this behavior to just yielding focus without clearing the input. There are some special case text fields that I have not touched (player name, map chooser, in-game chat). In the future we may set this behavior as the default for text inputs so that chrome logic doesn't need to implement it.

Closes #15426

pchote commented:
On macOS at least its a standard UI convention to be able to use <kbd>Esc</kbd> to clear search fields. I'm not necessarily opposed to removing this, but i wonder if we should think about adding some other quick mechanism to clear search fields? macOS also adds a floating :negative_squared_cross_mark: right aligned inside the field once there is text.

dragunoff commented:
> On macOS at least its a standard UI convention to be able to use Esc to clear search fields. I'm not necessarily opposed to removing this, but i wonder if we should think about adding some other quick mechanism to clear search fields? macOS also adds a floating negative_squared_cross_mark right aligned inside the field once there is text.

I don't mind changing the behavior (Esc clears, then yields if empty) as long as it's consistent across fields.

pchote commented:
That sounds best to me (for search fields only). We could add a search glyph in the left of the field if we want to visually distinguish them from other types of text field.

dragunoff commented:
> That sounds best to me (for search fields only). We could add a search glyph in the left of the field if we want to visually distinguish them from other types of text field.

What about the chat input then? It's not a search field and I want to standardize its behavior too.

dragunoff committed:
Make text fields yield keyboard focus on "Esc" in a consistent way

- search fields clear the input and yield if empty
- chat field and actor edit field yield without clearing

dragunoff commented:
Made some amendments as discussed:
- search fields clear the input and yield if empty
- chat field and actor edit field yield without clearing

pchote commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -133,7 +133,15 @@ public AssetBrowserLogic(Widget widget, Action onExit, ModData modData, WorldRen
+				if (string.IsNullOrEmpty(filenameInput.Text))
+					filenameInput.YieldKeyboardFocus();
+				else
+					filenameInput.Text = "";
```
Also need to call `ApplyFilter` to actually reset this.

dragunoff replied:
How about `filenameInput.OnTextEdited()` (it calls `ApplyFilter`)?

pchote replied:
Sure

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Editor/CommonSelectorLogic.cs:
```diff
@@ -46,8 +46,11 @@ public CommonSelectorLogic(Widget widget, World world, WorldRenderer worldRender
+				if (string.IsNullOrEmpty(SearchTextField.Text))
+					SearchTextField.YieldKeyboardFocus();
+				else
+					SearchTextField.Text = "";
```
And `SearchTextField.OnTextEdited()` here.

pchote reviewed (commented):
LGTM otherwise

dragunoff commented:
Update: Addressed review comments.

abcdefg30 reviewed (approved):
Ok.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e30e8ac389c0f637828690802c800715f679cb72...dd2f05a00e66dc23dd8a7e713755dec8a5f01eb8)
